plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/nessie/NessieIcebergClient.java

```diff
@@ -0,0 +1,408 @@
+
+    public UpdateableReference getReference()
+    {
+        return reference.get();
+    }
+
+    public List<SchemaTableName> listTables(Optional<String> namespace)
+    {
+        try {
+            return nessieApi.getEntries().reference(getReference().getReference())
+                    .get().getEntries().stream()
+                    .filter(entry -> namespace.map(ns -> Namespace.parse(ns).equals(entry.getName().getNamespace())).orElse(true))
+                    .filter(entry -> entry.getType() == Content.Type.ICEBERG_TABLE)
+                    .map(entry -> new SchemaTableName(entry.getName().getNamespace().name(), entry.getName().getName()))
+                    .collect(toImmutableList());
+        }
+        catch (NessieNotFoundException e) {
+            throw new TrinoException(ICEBERG_CATALOG_ERROR, e);
+        }
+    }
+
+    @Nullable
+    public IcebergTable loadTable(SchemaTableName schemaTableName)
+    {
+        try {
+            ContentKey key = NessieIcebergUtil.toKey(schemaTableName);
+            Content table = nessieApi.getContent().key(key).reference(getReference().getReference())
+                    .get().get(key);
+            return table != null ? table.unwrap(IcebergTable.class).orElse(null) : null;
+        }
+        catch (NessieNotFoundException e) {
+            return null;
+        }
+    }
+
+    private UpdateableReference loadReference(Optional<String> requestedRef, String hash)
+    {
+        try {
+            Reference ref = requestedRef.isEmpty() ? nessieApi.getDefaultBranch()
+                    : nessieApi.getReference().refName(requestedRef.get()).get();
+            // hash is currently always null, but a user will eventually be able to specify a particular branch/tag with a given hash to operate on
+            if (hash != null) {
+                if (ref instanceof Branch) {
+                    ref = Branch.of(ref.getName(), hash);
+                }
+                else {
+                    ref = Tag.of(ref.getName(), hash);
+                }
+            }
+            return new UpdateableReference(ref, hash != null);
+        }
+        catch (NessieNotFoundException e) {
+            if (requestedRef.isPresent()) {
+                throw new TrinoException(ICEBERG_CATALOG_ERROR, format("Nessie ref '%s' does not exist. This ref must exist before creating a NessieCatalog.", requestedRef), e);
+            }
+
+            throw new TrinoException(ICEBERG_CATALOG_ERROR, "Nessie does not have an existing default branch." +
+                    "Either configure an alternative ref via 'iceberg.nessie.ref' or create the default branch on the server.", e);
+        }
+    }
+
+    public void refreshReference()
+    {
+        try {
+            getReference().refresh();
+        }
+        catch (NessieNotFoundException e) {
+            throw new TrinoException(ICEBERG_CATALOG_ERROR, format("Failed to refresh as ref '%s' is no longer valid.", getReference().getName()), e);
```

Comment:
if `getReference()` throws an exception, also within the throw clause will be thrown another exception (because `getReference()` is called yet again).

Reply:
in this case the exception would only happen during `refresh`, so in this case it's safe to call `getReference().getName()`